### PR TITLE
cluster manager: Add option to create pool per downstream connection

### PR DESCRIPTION
**Pull request by chradcliffe**
Commit Message:
* Add connect_pool_per_downstream_connection flag to the cluster config (disabled by default)
* Add a `hashKey` method to Connection in order to be able to hash on the connection ID

Additional Description:
This is one part of the resolution for #12370 that allows for the downstream connection ID to be used as part of the hash key for the connection pool lookup.

Risk Level: Low (disabled by default)
Testing:
* Unit tests
* Manual testing to validate that a new pool is created for each downstream connection only when the option is enabled

Docs Changes: API comment
Release Notes: Updated for new flag

### CODE REVIEW

File: api/envoy/config/cluster/v3/cluster.proto [hunk 1]
```diff
@@ -917,6 +917,10 @@ message Cluster {
   // [#not-implemented-hide:]
   // Prefetch configuration for this cluster.
   PrefetchPolicy prefetch_policy = 50;
+
+  // If `connection_pool_per_downstream_connection` is true, the cluster will use a separate
+  // connection pool for every downstream connection
+  bool connection_pool_per_downstream_connection = 51;
```
Comment (htuch) [inline api/envoy/config/cluster/v3/cluster.proto:923]: Are you planning on adding release notes when your PR series wraps, or is it valid to add them now?

Comment (chradcliffe) [reply api/envoy/config/cluster/v3/cluster.proto:923]: Now that you mention it, it probably makes sense to put the release notes in with this PR as the follow-up is fairly separate.

File: source/common/network/connection_impl_base.cc [hunk 2]
```diff
@@ -3,13 +3,26 @@
 namespace Envoy {
 namespace Network {
 
+void ConnectionImplBaseUtility::addIdToHashKey(std::vector<uint8_t>& hash_key,
+                                               uint64_t connection_id) {
+  // Pack the connection_id into sizeof(connection_id) uint8_t entries in the hash_key vector
```
Comment (snowp) [inline source/common/network/connection_impl_base.cc:8]: nit: end comments with periods

File: source/common/network/connection_impl_base.h [hunk 3]
```diff
@@ -8,6 +8,19 @@
 namespace Envoy {
 namespace Network {
 
+/**
+ * Utility functions for the connection implementation base class.
+ */
+class ConnectionImplBaseUtility {
```
Comment (snowp) [inline source/common/network/connection_impl_base.h:14]: I'm not sure if an additional class is adding any value, did you consider just adding this as a static function on `ConnectionImplBase` itself?

File: test/common/network/connection_impl_test.cc [hunk 4]
```diff
@@ -75,6 +75,25 @@ TEST(ConnectionImplUtility, updateBufferStats) {
...
+TEST(ConnectionImplBaseUtility, addIdToHashKey) {
+  uint64_t connection_id = 0x0123456789abcdef;
+  std::vector<uint8_t> hash{{255, 254, 253, 252}};
+  ConnectionImplBaseUtility::addIdToHashKey(hash, connection_id);
+  ASSERT_EQ(12, hash.size());
+  EXPECT_EQ(255, hash[0]);
+  EXPECT_EQ(254, hash[1]);
+  EXPECT_EQ(253, hash[2]);
+  EXPECT_EQ(252, hash[3]);
+  EXPECT_EQ(0xef, hash[4]);
```
Comment (snowp) [inline test/common/network/connection_impl_test.cc:87]: Maybe use 0x consistently throughout?

File: source/common/network/connection_impl_base.h (hunk 3 above)
Comment (chradcliffe) [reply source/common/network/connection_impl_base.h:14]: My intention was to be consistent with [`ConnectionImplUtility`](https://github.com/envoyproxy/envoy/blob/master/source/common/network/connection_impl.h#L39) but I can also just make it a static function.

**Pull request merged**